### ISSUE TITLE
Getting field names

### ISSUE THREAD

**Issue by 3lnc** [elastic/elasticsearch-dsl-py#144]
TL;DR: Is there any "simple" way to get document field names?

Despite ES being schemaless, I believe that many people works with ES respecting schema (which is pretty natural, as we can use object mapping). Personally, I try to verify any data on app side, to avoid implicit mapping, which can lead to troubles in future indexing/migrations.
And as ES have no "constraints" against dynamic mapping and `elasticsearch-dsl-py` allows to save docs via DocType(`**data`) unpacking, I've extended `DocType` with
```python
@classmethod
    def get_field_names(cls, path=None):
        if path is None:
            return cls._doc_type.mapping.properties.properties._d_.keys()
        else:
            getter = attrgetter('.'.join('{}.properties'.format(f) for f in path.split('.') if f))
            try:
                return getter(cls._doc_type.mapping.properties.properties)._d_.keys()
            except AttributeError:
                return []
```
which acts like Django's `get_all_field_names`, but also allows to get field names for `Nested` and `Object` fields. Is there a simpler way, or such usecase is pretty rare?
Thx.

**Comment by HonzaKral** [comment]
ES isn't really schema-less, that is a common misconception based on it having a dynamic schema. It does have one. Also it has constraints against dynamic mapping - you can set dynamic to `false` (causing es to ignore unmapped fields) or you can set it to `strict` (causing ES to raise an error on encountering unmapped field) [0]. This made me realize that we don't have a way to set this in the python dsl (created #146 to track.)

With that information I don't think we need a `get_field_by_name`, or `get_fields`. But if you think it would be useful and others agree we can definitely add these method to `DocTypeOptions` where I think these should live.

Does this make sense?

Thanks

0 - http://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-dynamic-mapping.html

**Comment by 3lnc** [comment]
Yea, I know that ES is schema-intensive with dynamic mapping ability.
Basically, that example is only thing I've faced (need to "validate" data for couple of doc types).

The proposal is to give ability to get fields/field names easily to give ability to simplify validation/code inspection if needed. Because now the chaining properties (including private) to get such seems terrible for me.

For sure, this is point to discuss.

**Comment by HonzaKral** [comment]
I agree that it's terrible, partly by design. But I would much rather fix the underlying cause - for example I do have plans to introduce proper validation logic that could be attached to fields (including `Nested` and `Object` of course) as well as entire `DocType` - both simple configuration (`required=True`) and more complex stuff (`def clean(self): if I_DONT_LIKE_DATA: raise ValidationError()`).

Does that make more sense?

**Comment by 3lnc** [comment]
Yeah, seems so.

**Comment by HonzaKral** [comment]
This has been resolved by e1bf242
